### ISSUE TITLE
Examples/Templates:  Some Roadblocks to Creating Pipes with Immediate Impact

### ISSUE THREAD

**Issue by wanderslth** [flexiodata/flexio-web-app#228]
tl;dr  These are thoughts from a deep dive in examples I did on Tuesday.  The questions is how do we make example pipes that are concrete, immediate and launch people into thinking about what they can build for themselves?  It's a tough nut to crack, but worth cracking.

Ben's thumbnail pipe example is so good because it allows a user to quickly validate an example with their own easy-to-access file.  The "paste-in-a-URL-to-get-a-word-cloud" example is similar in that someone can immediately test with their own url and then see how the pipe made the magic happen.  However, it is difficult to replicate this type of parlor trick with the bulk of Flex's functionality, since tables are so fussy.

...

I was attempting various template ideas for the onboarding and I've run into a bunch of blind alley's, mostly because 95% of flex operations require field names, which are unknown unless you have a hard-coded file.  In other words, users can't easily swap in their own data to try out a template.  

Sadly, this means they're effectively running a sample pipe, which probably isn't going to provide the dopamine kick of "look, my operations worked and transformed this" or "look, my file is now transformed and moved to this".

For posterity, here are a few of the alleys:

* **Table-based operations.**  Since almost all functionality requires field names, its difficult to have people swap in their own random CSV file without a lot of explanation and variables.  `convert header: ${value} qualifier: ${value}` or `transform col: ${value(s?)} case: upper` or `some-kind-of-type-conversion` etc.

* **File Locations.**  Given the above, it doesn't matter if we had a file upload or easy way for them to switch to a google sheet, etc., unless the example is simply a 'move' or 'conversion'.  That said, if we did have options in addition to "table-based", then we do run into the issue that their data/file needs to be web-based to try (ie, no upload).

* **Multiple Files.**  There is no non-connection way I can see of providing a sample folder with, say, 3 pngs and 3 csvs and then have them swap out to their own connection if they want.  Therefore, you have to rely on explaining to the user the the kind of folder they need and hope they have one handy to, say, run an operation where you filter for `*.csv` or `fileprefix*.csv`   (and, for the latter, you need hard-coded filenames or very busy templates)

* **Text/JSON based operations.**  We currently have three operations that work on text:  `limit`, `merge` and `transform case:`  `merge` requires multiple files (see above).  `limit` and `transform case:` are not terribly soul-stirring. Python is basically required to make non-table processing interesting.

* **PDF/Images.**  It's very cool that our package can convert PDF to text, but then you run into what you might actually do with it when you have it (see above).  Again, python is required for anything beyond `move/convert`.

---

Thoughts:

* **Table operations.**  I'm not against hard-coded examples, which is what we are doing now.  In fact, these can provide super helpful recipes for people doing table/ETL kinda work (like our man jonathan).  There can be 100s of these with calculated fields alone that do the various data tricks we're all familiar with.  But, in the medium-term, they probably aren't best suited for the "top 3 templates" you onboard people with.

* **Python.**  I have no problem heavily leveraging python.  But, it will limit any non-python users.  And it will force the bulk of example building to Aaron/Ben.

* **API.**  We're going to be pushing the API angle harder, but these are the most difficult to "show" via an example on its own sans an external web page, etc.  So, need to think through how to best use in conjunction with additional tutorials/external static sites.

* **Field Agnostic Operations.**  We've talked about these in the past, particularly with JSON related things, validation steps, multi-text file processing, etc..  Given the CSV-to-JSON interest we got at U of C, maybe not a terrible thing to look at again if there is low-hanging fruit somewhere.

* **Multi-file repos.**   If we had a public github repo (read-only) or a public google drive folder (read-only) maybe we could make it easier to have people swap in their own connections to do multi-file processing.  It would be nice to have something work out of the box though without making people set up a connection and find our public store... any ideas?

Long and short, another thing we should be thinking through to help translate our abstract pipes into something more immediate.
